### PR TITLE
statistics: support hotPeerCache maintained in Peer level

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: yisaer <disxiaofei@163.com>

### What is changed and how it works?

make hotPeerCache maintained by peer level

### Release note

* No release note

### CODE REVIEW

File: server/statistics/hot_peer_cache.go [hunk 1]
```diff
@@ -209,6 +170,58 @@ func (f *hotPeerCache) CheckRegionFlow(region *core.RegionInfo) (ret []*HotPeerS
...
 
+// CheckPeerFlow checks the flow information of a peer.
+func (f *hotPeerCache) CheckPeerFlow(peer *core.PeerInfo, region *core.RegionInfo, interval uint64) *HotPeerStat {
+	storeID := peer.GetStoreID()
+	bytes := float64(f.getPeerBytes(peer))
+	keys := float64(f.getPeerKeys(peer))
+	byteRate := bytes / float64(interval)
+	keyRate := keys / float64(interval)
+	f.collectPeerMetrics(byteRate, keyRate, interval)
+	var tmpItem *HotPeerStat
```
Comment (lhy1024) [inline server/statistics/hot_peer_cache.go:181]: > it may has been moved to other store, we save it to tmpItem. 

Putting it into CheckPeerFlow does not achieve this goal

File: server/statistics/hot_peer_cache.go [hunk 2]
```diff
@@ -450,3 +437,70 @@ func (f *hotPeerCache) updateHotPeerStat(newItem, oldItem *HotPeerStat, bytes, k
 	}
 	return newItem
 }
+
+// TODO: remove it in future
+func (f *hotPeerCache) getRegionBytes(region *core.RegionInfo) uint64 {
```
Comment (lhy1024) [inline server/statistics/hot_peer_cache.go:442]: It seems to be related with #3591

File: server/statistics/hot_peer_cache.go [hunk 3]
```diff
@@ -454,3 +412,93 @@ func (f *hotPeerCache) updateHotPeerStat(newItem, oldItem *HotPeerStat, deltaLoa
...
+		case RegionWriteBytes:
+			writeByteHist.Observe(loads[int(k)])
+		case RegionWriteKeys:
+			writeKeyHist.Observe(loads[int(k)])
+		}
+	}
+}
+
+// gets the storeIDs, including old region and new region
+func (f *hotPeerCache) getAllStoreIDs(region *core.RegionInfo) []uint64 {
```
Comment (lhy1024) [inline server/statistics/hot_peer_cache.go:438]: What is the difference between it and the previous one?

File: server/core/peer.go [hunk 4]
```diff
@@ -72,3 +72,75 @@ func CountInJointState(peers ...*metapb.Peer) int {
 	}
 	return count
 }
+
+// PeerInfo provides peer information
+type PeerInfo struct {
+	peerID       uint64
+	StoreID      uint64
```
Comment (rleungx) [inline server/core/peer.go:79]: Why public?

File: server/statistics/hot_peer_cache.go (hunk 3 above)
Comment (Yisaer) [reply server/statistics/hot_peer_cache.go:438]: no difference

File: server/statistics/hot_peer_cache.go [hunk 5]
```diff
@@ -224,6 +161,51 @@ func (f *hotPeerCache) CheckRegionFlow(region *core.RegionInfo) (ret []*HotPeerS
...
 
+// CheckPeerFlow checks the flow information of a peer.
+func (f *hotPeerCache) CheckPeerFlow(peer *core.PeerInfo, region *core.RegionInfo, interval uint64) *HotPeerStat {
+	storeID := peer.GetStoreID()
+	deltaLoads := f.getPeerDeltaLoads(peer)
+	loads := make([]float64, len(deltaLoads))
+	for i := range deltaLoads {
+		loads[i] = deltaLoads[i] / float64(interval)
+	}
+	f.collectPeerMetrics(loads, interval)
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply server/statistics/hot_peer_cache.go:172]: removed.

File: server/core/peer.go (hunk 4 above)
Comment (Yisaer) [reply server/core/peer.go]: updated.

File: server/core/peer.go [hunk 6]
```diff
@@ -72,3 +72,75 @@ func CountInJointState(peers ...*metapb.Peer) int {
 	}
 	return count
 }
+
+// PeerInfo provides peer information
+type PeerInfo struct {
+	peerID       uint64
```
Comment (rleungx) [inline server/core/peer.go:78]: Why not directly use metapb.Peer?

Comment (Yisaer) [reply server/core/peer.go:78]: We use `PeerInfo` to maintain the information like `readKeys` / `readBytes`.

Comment (rleungx) [reply server/core/peer.go:78]: I mean
```
type PeerInfo struct {
	peer           *metapb.Peer
	readKeys       uint64
        ...
}
```

Comment (Yisaer) [reply server/core/peer.go:78]: That's ok to me.

Comment (rleungx) [reply server/core/peer.go:78]: We don't need `FromMetaPeer` anymore

File: server/core/peer.go [hunk 7]
```diff
@@ -72,3 +72,66 @@ func CountInJointState(peers ...*metapb.Peer) int {
...
 	return count
 }
+
+// PeerInfo provides peer information
+type PeerInfo struct {
+	*metapb.Peer
+	writtenBytes uint64
+	writtenKeys  uint64
+	readBytes    uint64
+	readKeys     uint64
```
Comment (lhy1024) [inline server/core/peer.go:82]: would you like to replace them with []float64?

File: server/statistics/hot_peer_cache.go [hunk 8]
```diff
@@ -108,120 +112,93 @@ func (f *hotPeerCache) Update(item *HotPeerStat) {
...
+		justTransferLeader: justTransferLeader,
+		interval:           interval,
+		peers:              peers,
+		thresholds:         thresholds,
+	}
+	if oldItem == nil {
+		for _, storeID := range f.getAllStoreIDs(region) {
+			oldItem = f.getOldHotPeerStat(region.GetID(), storeID)
+			if oldItem != nil {
+				break
```
Comment (lhy1024) [inline server/statistics/hot_peer_cache.go:197]: The behavior is different. The previous code ensured that it inherited exactly the right information by saving the expired item as a tmpItem. The current code is equivalent to inheriting a random item.

File: server/statistics/hot_peer_cache.go [hunk 9]
```diff
@@ -454,3 +435,47 @@ func (f *hotPeerCache) updateHotPeerStat(newItem, oldItem *HotPeerStat, deltaLoa
 	}
 	return newItem
 }
+
+// TODO: remove it in future
+func (f *hotPeerCache) collectRegionMetrics(loads []float64, interval uint64) {
```
Comment (lhy1024) [inline server/statistics/hot_peer_cache.go:440]: Do you mind putting this function in the previous location? Changing the position brings unnecessary diff

File: server/statistics/hot_peer_cache.go [hunk 10]
```diff
@@ -255,10 +232,12 @@ func (f *hotPeerCache) getOldHotPeerStat(regionID, storeID uint64) *HotPeerStat
 	return nil
 }
 
-func (f *hotPeerCache) isRegionExpired(region *core.RegionInfo, storeID uint64) bool {
+func (f *hotPeerCache) isPeerExpired(peer *core.PeerInfo, region *core.RegionInfo) bool {
+	storeID := peer.GetStoreID()
 	switch f.kind {
 	case WriteFlow:
 		return region.GetStorePeer(storeID) == nil
```
Comment (rleungx) [inline server/statistics/hot_peer_cache.go:239]: Seem it is unreachable?

File: server/core/peer.go (hunk 7 above)
Comment (Yisaer) [reply server/core/peer.go:82]: We will change it with regionInfo together in future.

File: server/statistics/hot_peer_cache.go (hunk 8 above)
Comment (Yisaer) [reply server/statistics/hot_peer_cache.go]: Currently, we tend to inherit stat from leader, otherwise we inherit stat from random peer. In future, we will use a new way to inherit stat more currectly.

File: server/statistics/hot_peer_cache.go [hunk 11]
```diff
@@ -224,6 +180,59 @@ func (f *hotPeerCache) CheckRegionFlow(region *core.RegionInfo) (ret []*HotPeerS
...
+	deltaLoads := f.getFlowDeltaLoads(peer)
+	loads := make([]float64, len(deltaLoads))
+	for i := range deltaLoads {
+		loads[i] = deltaLoads[i] / float64(interval)
+	}
+	justTransferLeader := f.justTransferLeader(region)
+	// transfer read leader or remove write peer
+	isExpired := f.isPeerExpired(peer, region)
+	oldItem := f.getOldHotPeerStat(region.GetID(), storeID)
+	if isExpired && oldItem != nil && !f.isInheritExist(region.GetID()) {
```
Comment (lhy1024) [inline server/statistics/hot_peer_cache.go:196]: why need to judge whether exist in inherit items?

Comment (Yisaer) [reply server/statistics/hot_peer_cache.go]: removed

File: server/statistics/hot_peer_cache.go [hunk 12]
```diff
@@ -454,3 +467,48 @@ func (f *hotPeerCache) updateHotPeerStat(newItem, oldItem *HotPeerStat, deltaLoa
...
+		case RegionWriteKeys:
+			ret[k] = float64(stat.GetKeysWritten())
+		}
+	}
+	return ret
+}
+
+func (f *hotPeerCache) isInheritExist(regionID uint64) bool {
+	item, ok := f.inheritItem[regionID]
+	return ok && item != nil
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply server/statistics/hot_peer_cache.go]: when there is no inherit item.

File: server/statistics/hot_cache.go [hunk 13]
```diff
@@ -121,3 +143,37 @@ func (w *HotCache) GetFilledPeriod(kind FlowKind) int {
 	}
 	return 0
 }
+
+func (w *HotCache) updateReadItems(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			close(w.readFlowQueue)
```
Comment (rleungx) [inline server/statistics/hot_cache.go:151]: It's better to close it on send side.

Comment (Yisaer) [reply server/statistics/hot_cache.go]: updated.

**Review by lhy1024 — APPROVED** [review]

File: server/cluster/cluster.go [hunk 14]
```diff
@@ -546,8 +546,8 @@ func (c *RaftCluster) processRegionHeartbeat(region *core.RegionInfo) error {
 		c.RUnlock()
 		return err
 	}
-	writeItems := c.CheckWriteStatus(region)
-	readItems := c.CheckReadStatus(region)
+	c.CheckWriteStatus(region)
```
Comment (rleungx) [inline server/cluster/cluster.go:549]: no need to use two functions

File: server/statistics/hot_cache.go [hunk 15]
```diff
@@ -121,3 +143,35 @@ func (w *HotCache) GetFilledPeriod(kind FlowKind) int {
...
+				items := w.readFlow.CheckRegionFlow(region)
+				for _, item := range items {
+					w.Update(item)
+				}
+			}
+		}
+	}
+}
+
+func (w *HotCache) updateWriteItems(ctx context.Context) {
```
Comment (rleungx) [inline server/statistics/hot_cache.go:163]: Can we use one goroutine?

File: server/statistics/hot_peer_cache.go [hunk 16]
```diff
@@ -454,3 +467,43 @@ func (f *hotPeerCache) updateHotPeerStat(newItem, oldItem *HotPeerStat, deltaLoa
...
+	f.inheritItem[item.RegionID] = item
+}
+
+func (f *hotPeerCache) takeInheritItem(regionID uint64) *HotPeerStat {
+	item, ok := f.inheritItem[regionID]
+	if !ok {
+		return nil
+	}
+	if item != nil {
+		ret := *item
```
Comment (rleungx) [inline server/statistics/hot_peer_cache.go:504]: Why not directly return item?

**Review by rleungx — APPROVED** [review]
